python/tskit/trees.py

```diff
@@ -4244,13 +4135,136 @@ def f2(self, sample_sets, indexes=None, windows=None, mode="site",
             self._ll_tree_sequence.f2, 2, sample_sets, indexes=indexes, windows=windows,
             mode=mode, span_normalise=span_normalise)
 
+    def mean_descendants(self, reference_sets):
+        """
+        Computes for every node the mean number of samples in each of the
+        `reference_sets` that descend from that node, averaged over the
+        portions of the genome for which the node is ancestral to *any* sample.
+        The output is an array, `C[node, j]`, which reports the total span of
+        all genomes in `reference_sets[j]` that inherit from `node`, divided by
+        the total span of the genome on which `node` is an ancestor to any
+        sample in the tree sequence.
+
+        .. note:: This interface *may change*, particularly the normalization by
+            proportion of the genome that `node` is an ancestor to anyone.
+
+        :param iterable reference_sets: A list of lists of node IDs.
+        :return: An array with dimensions (number of nodes in the tree sequence,
+            number of reference sets)
+        """
+        # An array of booleans per tree, true if node u is in the tree else false.
+        in_tree = self.sample_count_stat(
+            [self.samples()], lambda x: x + 1, output_dim=1, polarised=True,
+            strict=False, mode="node", windows="trees")[..., 0] > 1
+        breakpoints = self.breakpoints(as_array=True)
+        tree_span = breakpoints[1:] - breakpoints[:-1]
+        tree_span = tree_span.reshape((tree_span.shape[0], 1))
+        # The total span over which each node is in a tree.
+        node_span = np.sum(in_tree * tree_span, axis=0)
+
+        C = self.sample_count_stat(
+            reference_sets, lambda x: x, output_dim=len(reference_sets), polarised=True,
+            strict=False, mode="node", windows="trees")
+        # Reshape so that we can multiply through by the span of each tree
+        tree_span = tree_span.reshape((tree_span.shape[0], 1, 1))
+        C *= tree_span
+
+        # To replicate the existing code, take the mean weighted by node span.
+        C = np.sum(C, axis=0)
```

Comment:
I think the following does the same thing without having to do the `windows="trees"` stuff yourself:
```
denom = self.sample_count_stat(
            [self.samples()], lambda x: x > 0, output_dim=1, polarised=True,
            strict=False, mode="node", span_normalise=False)
numer = self.sample_count_stat(
        reference_sets, lambda x: x, output_dim=len(reference_sets), polarised=True, span_normalise=False,
        strict=False, mode="node")
C = numer / denom
```

Reply:
This sort of normalization by `denom` is a pretty common node statistic operation that we should maybe even define as an option to those methods.

Reply:
Or maybe (as we had said on this topic before) just provide a method that gives us the denominator. As a site statistic this'd give the number of segregating-or-fixed sites, and as a branch statistic would give us the total tree length plus any branches above the MRCA. So maybe it should be an option to `segregating_sites`, say, `include_fixed=False`.